twill-yarn/src/main/java/org/apache/twill/internal/container/TwillContainerService.java

```diff
@@ -51,29 +63,36 @@
   private final ClassLoader classLoader;
   private final BasicTwillContext context;
   private final ContainerLiveNodeData containerLiveNodeData;
+  private final Map<String, LogEntry.Level> oldLogLevels;
+  private final Map<String, LogEntry.Level> defaultLogLevels;
   private ExecutorService commandExecutor;
   private TwillRunnable runnable;
 
   public TwillContainerService(BasicTwillContext context, ContainerInfo containerInfo, ZKClient zkClient,
                                RunId runId, TwillRunnableSpecification specification, ClassLoader classLoader,
-                               Location applicationLocation) {
+                               Location applicationLocation, Map<String, LogEntry.Level> defaultLogLevels,
+                               Map<String, LogEntry.Level> logLevels) {
     super(zkClient, runId, applicationLocation);
 
     this.specification = specification;
     this.classLoader = classLoader;
-    this.containerLiveNodeData = createLiveNodeData(containerInfo);
+    this.defaultLogLevels = ImmutableMap.copyOf(defaultLogLevels);
+    this.oldLogLevels = Collections.synchronizedMap(new HashMap<>(defaultLogLevels));
```

Comment:
The `defaultLogLevels` are the log levels passed with `TwillPreparer`, when resetting the log levels, we will need it to update the live node info to indicate the log levels. The `oldLogLevels` is the same as the `defaultLogLevels` when the runnable starts up. We will not update it for each log level change. It is used to save old log levels for loggers for the first time change as we do not know what is in the `logback.xml`. When reset, we reset based on this map.